api/src/main/java/org/openmrs/module/queue/api/impl/QueueEntryServiceImpl.java

```diff
@@ -131,6 +131,30 @@ public QueueEntry transitionQueueEntry(QueueEntryTransition queueEntryTransition
 		return getProxiedQueueEntryService().saveQueueEntry(queueEntryToStart);
 	}
 	
+	/**
+	 * @see QueueEntryService#undoTransition(QueueEntry)
+	 */
+	@Override
+	public QueueEntry undoTransition(@NotNull QueueEntry queueEntry) {
+		if (queueEntry.getVoided()) {
+			throw new IllegalArgumentException("cannot undo transition on a voided queue entry");
+		}
+		if (queueEntry.getEndedAt() != null) {
+			throw new IllegalArgumentException("cannot undo transition on an ended queue entry");
+		}
+		QueueEntry prevQueueEntry = getPreviousQueueEntry(queueEntry);
+		if (prevQueueEntry == null) {
+			throw new IllegalArgumentException("specified queue entry does not have a previous queue entry");
+		}
```

Comment:
All 3 of the above exceptions would be best to throw in such a way that the UI can present them to the user in a translatable fashion, and one in which the message can be customized at the UI-level, not hard-coded on the backend level.  So it would be best to have some sort of code in the Exception that can be passed through to any error object returned to the frontend via the REST response.